Delete segment completely when ZkOperator fails to update IdealState with ZkInterruptedException

AlexanderKM opened the pull request:
For https://github.com/apache/pinot/issues/16866

This PR adds a segment deletion step to the cleanup process when a new segment is uploaded and we encounter a `ZkInterruptedException` while updating the IdealState. Ultimately because of this exception, we cannot be certain if the IdealState was updated successfully, and since the code will already delete the segment metadata from the propertystore, it is a good idea to delete it from the IdealState as well.

We have been using this patch in our deployment for some months and it has been helpful in preemptively catching this scenario so we don't need to go in and manually delete the segment ourselves.

AlexanderKM committed:
Delete segment completely when ZkOperator fails to update IdealState with ZkInterruptedException

Signed-off-by: Alex Maniates <maniatesa@gmail.com>

Jackie-Jiang reviewed (commented):
Thanks for contributing the fix!

Is there a way to identify if an IS update is persisted? Currently we assume the update not persisted when exception is thrown, which caused the issue. That is the root problem to fix

AlexanderKM committed:
Move segment deletion cleanup into PinotHelixResourceManager

Signed-off-by: Alex Maniates <maniatesa@gmail.com>

Jackie-Jiang commented:
We should also consider retrying for such temporary exceptions.
Clean up is fine, but what if the clean up also fails?

AlexanderKM committed:
Refactor PinotHelixResourceManager to retry segment assignment when encountering ZkInterruptedException

Signed-off-by: Alex Maniates <maniatesa@gmail.com>

AlexanderKM commented:
I took your advice and added a retry in the segment assignment flow! Just only when a ZkInterruptedException is encountered (it probably doesn't make sense to retry in other scenarios since the [RetryPolicy](https://github.com/apache/pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/common/utils/helix/IdealStateGroupCommit.java#L209-L211) is already attempting 20 times in "normal" cases, so keeping this scoped to only ZK exceptions for now). This did require a bit of refactoring but I think it looks better in the end.

To support this, I added some unit tests for the following:
- Successful add new segment
- Recovering from ZkInterruptedException leads to a successful add new segment
- Complete failure of adding new segment in case of multiple ZK exceptions or other normal exceptions like AttemptsExceeded

>Clean up is fine, but what if the clean up also fails?

I think there will inevitably be some scenarios that we cannot completely recover from, but we will at least let the caller know that an error occurred and then they can handle cleanup manually as needed. This also hopefully removes the segment from both the PropertyStore AND IdealState, which leaves the zookeeper state in a better place 🤞 

Let me know your thoughts when you have time @Jackie-Jiang

AlexanderKM commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:
```diff
@@ -2552,51 +2554,117 @@ public List<Tier> getSortedTiers(TableConfig tableConfig) {
+    HelixHelper.updateIdealState(_helixZkManager, tableNameWithType, idealState -> {
+      assert idealState != null;
+      Map<String, Map<String, String>> currentAssignment = idealState.getRecord().getMapFields();
+      if (currentAssignment.containsKey(segmentName)) {
```
Also note: by retrying, we are also checking if the segment exists automatically here which is nice

Jackie-Jiang commented:
Digging more into the code, I'm still trying to figure out when will it get `ZkInterruptedException` but with IS persisted. So far I have found Helix `ZkClient.writeDataReturnStat()` -> `ZkClient.retryUntilConnected()` could throw `ZkInterruptedException` but it shouldn't persist the IS.
The fix in this PR is not general enough because the core problem is IS update result is no longer consistent, which could impact a lot of logics, in addition to segment upload. We need to fix the IS update logic to ensure it either successfully persist the result, or throw exception.
@xiangfu0 Can you help take a look at this? Can we handle this in `IdealStateGroupCommit`?

Jackie-Jiang commented:
@AlexanderKM Great finding of the old Helix issue. I cannot re-open it, but opened a new one here: https://github.com/apache/helix/issues/3072
